### PR TITLE
Stop exporting `TypeckRootCtxt` and `FnCtxt`.

### PR DESCRIPTION
**Pull request by oli-obk**
While they have many convenient APIs, it is better to expose dedicated functions for them

noticed in #122213

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 1]
```diff
@@ -60,8 +62,6 @@ pub struct CastCheck<'tcx> {
     cast_ty: Ty<'tcx>,
     cast_span: Span,
     span: Span,
-    /// whether the cast is made in a const context or not.
-    pub constness: hir::Constness,
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/cast.rs:64]: Also helps with the compiler telling us about unused things

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 2]
```diff
@@ -194,18 +194,47 @@ fn make_invalid_casting_error<'a, 'tcx>(
     )
 }
 
+/// If a cast from `from_ty` to `to_ty` is valid, returns an Ok containing the kind of
+/// the cast. In certain cases, including some invalid casts from array references
+/// to pointers, this may cause additional errors to be emitted and/or ICE error
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/cast.rs:199]:
```suggestion
/// If a cast from `from_ty` to `to_ty` is valid, returns a `Some` containing the kind
/// of the cast. In certain cases, including some invalid casts from array references
/// to pointers, this may cause additional errors to be emitted and/or ICE error
```

This returns an `Option`, not a `Result`.

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 3]
```diff
@@ -194,18 +194,47 @@ fn make_invalid_casting_error<'a, 'tcx>(
...
+) -> Option<CastKind> {
+    let hir_id = e.hir_id;
+    let local_def_id = hir_id.owner.def_id;
+
+    let root_ctxt = crate::TypeckRootCtxt::new(tcx, local_def_id);
+    let fn_ctxt = FnCtxt::new(&root_ctxt, param_env, local_def_id);
+
+    if let Ok(check) = CastCheck::new(
+        &fn_ctxt, e, from_ty, to_ty,
+        // We won't show any error to the user, so we don't care what the span is here.
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/cast.rs:218]:
```suggestion
        // We won't show any errors to the user, so the span is irrelevant here.
```

Would probably read better (I know you didn't write this but seems nice to do a drive-by)

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 4]
```diff
@@ -194,18 +194,47 @@ fn make_invalid_casting_error<'a, 'tcx>(
     )
 }
 
+/// If a cast from `from_ty` to `to_ty` is valid, returns an Ok containing the kind of
+/// the cast. In certain cases, including some invalid casts from array references
+/// to pointers, this may cause additional errors to be emitted and/or ICE error
+/// messages. This function will panic if that occurs.
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/cast.rs:200]: Is this really true? I don't see any panicking done in the body of this function, but I assume `CastCheck` wouldn't do that.

**Review by fee1-dead — CHANGES_REQUESTED** [review]
I hope you don't mind the comment nits. r=me afterwards :)

File: compiler/rustc_hir_typeck/src/cast.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/cast.rs:200]: Oh heh. I didn't read the comment at all and just copied over the function
